Add microchip lan867x revd0 phy support

ParthibanI17164 opened the pull request:
Summary

This PR adds support for the Microchip LAN8670/1/2 Rev.D0 10BASE-T1S PHYs and implements proper link status control for the new silicon revision.

The LAN867x Rev.D0 devices introduce updated initialization requirements and behavior changes that differ from previous revisions, as outlined in Microchip Application Note AN1699 (Revision G, DS60001699G – October 2025).
🔗 [Microchip AN1699 – 10BASE-T1S PHY Initialization and Configuration](https://www.microchip.com/en-us/application-notes/an1699)

Changes Included

1. Add LAN867x Rev.D0 PHY Support

- Introduces specific initialization sequences required for optimal operation and OPEN Alliance compliance.
- Ensures compatibility with Rev.D0 silicon as described in the latest application note.
- Maintains backward compatibility with earlier silicon revisions.

2. Configure Link Status Control for Rev.D0

- Implements proper link status handling for LAN867x Rev.D0 devices:
  - PLCA Mode: Link status reflects the PLCA operational state.
  - CSMA/CD Mode: Since autonegotiation is not supported, the link status is forced active using the LINK_STATUS_SEMAPHORE bit.
- Configuration occurs:
  - During PHY initialization (default CSMA/CD mode).
  - Whenever PLCA configuration is updated.
- Ensures consistent and accurate link reporting across modes.

ParthibanI17164 committed:
drivers: ethernet: phy: add support for Microchip LAN867X Rev.D0 PHY

Add support for the LAN8670/1/2 Rev.D0 10BASE-T1S PHYs from Microchip.
The new Rev.D0 silicon requires a specific set of initialization
settings to be configured for optimal performance and compliance with
OPEN Alliance specifications, as described in Microchip Application Note
AN1699 (Revision G, DS60001699G – October 2025).
https://www.microchip.com/en-us/application-notes/an1699

Signed-off-by: Parthiban Veerasooran <parthiban.veerasooran@microchip.com>

ParthibanI17164 committed:
drivers: ethernet: phy: configure link status control for LAN867x Rev.D0

Configure the link status in the Link Status Control register for
LAN8670/1/2 Rev.D0 PHYs, depending on whether PLCA or CSMA/CD mode
is enabled. When PLCA is enabled, the link status reflects the PLCA
status. When PLCA is disabled (CSMA/CD mode), the PHY does not support
autonegotiation, so the link status is forced active by setting
the LINK_STATUS_SEMAPHORE bit.

The link status control is configured:
- During PHY initialization, for default CSMA/CD mode.
- Whenever PLCA configuration is updated.

This ensures correct link reporting and consistent behavior for
LAN867x Rev.D0 devices.

Signed-off-by: Parthiban Veerasooran <parthiban.veerasooran@microchip.com>

maass-hamburg commented on drivers/ethernet/phy/phy_microchip_t1s.c:
```diff
@@ -80,6 +92,22 @@ static lan865x_config lan865x_revb_config[] = {
+ * https://www.microchip.com/en-us/application-notes/an1699
+ * Revision G (DS60001699G - October 2025)
+ */
+static lan865x_config lan867x_revd0_config[] = {
```
this seems to never change. add `const`?

ParthibanI17164 replied:
Sure, will do.

maass-hamburg commented on drivers/ethernet/phy/phy_microchip_t1s.c:
```diff
@@ -533,6 +612,13 @@ static int phy_mc_t1s_init(const struct device *dev)
+		if (ret) {
+			LOG_ERR("PHY initial configuration error: %d\n", ret);
+			return ret;
+		}
```
might be a good idea to move this part out of the switch case, as this is the same, for all cases except the default

ParthibanI17164 commented on drivers/ethernet/phy/phy_microchip_t1s.c:
```diff
@@ -533,6 +612,13 @@ static int phy_mc_t1s_init(const struct device *dev)
 		break;
+	case PHY_ID_LAN867X_REVD0:
+		ret = phy_mc_lan867x_revd0_config_init(dev);
+		if (ret) {
```
Ok will change in all places.

ParthibanI17164 committed:
drivers: ethernet: phy_microchip_t1s: use 'ret < 0' for error checks

Replace 'if (ret)' with 'if (ret < 0)' in the Microchip T1S PHY driver.
This change follows the legacy coding style commonly used in Zephyr
drivers, where error conditions are checked explicitly against negative
values. The affected functions do not return positive values, so this
modification does not change functionality. No functional change intended.

Signed-off-by: Parthiban Veerasooran <parthiban.veerasooran@microchip.com>

pdgendt reviewed (approved):
Mind you, we are in feature freeze, so this will go in after.
